wasm2c/wasm-rt-impl.c

```diff
@@ -68,6 +104,9 @@ uint32_t wasm_rt_saved_call_stack_depth;
 
 static FuncType* g_func_types;
 static uint32_t g_func_type_count;
+// Mutex for the func type indexes which are shared across multiple Wasm
+// instances
```

Comment:
", potentially running on different threads"

Reply:
Would a valid alternative simply be to make g_func_types and g_func_type_count into TLS?   I guess that would require just as much work since C11 `_Thread_local` keyword cannot be relied on here?

Reply:
> ", potentially running on different threads"
:+1: 

> Would a valid alternative simply be to make g_func_types and g_func_type_count into TLS?

Independent of how we could implement TLS, I think the bigger issue is that TLS wouldn't work here. This is because `g_func_types` and `g_func_type_count` must have the same value when queried from (different instances on) different threads

Reply:
Can you explain why instances need to share `g_func_types`?   Its it an internal detail?

Reply:
@sbc100 Actually, can we assume C11? If yes, I can probably use [C11 mutex support](https://en.cppreference.com/w/c/thread/mtx_lock). For what its worth, the next PR I plan to submit will be to make use of TLS vars for things like exception handling, stack depth vars etc. which is again much simpler if we have C11.

Reply:
> Can you explain why instances need to share g_func_types? Its it an internal detail?

@sbc100 I was mostly basing my reasoning on @keithw 's comment from [here](https://github.com/WebAssembly/wabt/issues/2046#issuecomment-1306638191)
Let me know if you believe this is not necessary, in which case, we can probably just move this to some sort of "module state" instead of global state

Reply:
> @sbc100 Actually, can we assume C11? If yes, I can probably use [C11 mutex support](https://en.cppreference.com/w/c/thread/mtx_lock). For what its worth, the next PR I plan to submit will be to make use of TLS vars for things like exception handling, stack depth vars etc. which is again much simpler if we have C11.

I think wasm2c is supposed to generate C89 and I assume that we would want to extend that to the rt impl.. so I think that answer is no, we should avoid anything outside of C89.

Reply:
> > Can you explain why instances need to share g_func_types? Its it an internal detail?
> 
> @sbc100 I was mostly basing my reasoning on @keithw 's comment from [here](https://github.com/WebAssembly/wabt/issues/2046#issuecomment-1306638191) Let me know if you believe this is not necessary, in which case, we can probably just move this to some sort of "module state" instead of global state

Maybe lets wait for Keith to chime in then.  I'm fine with this type of locking, or making types module internal.   Because this only happens during module init I think that locking should have minimal impact.

Reply:
Longer comment below. The (integer) func types have to be unique for different types across the context of "a module and the modules it transitively imports funcrefs and funcref tables from," because this is how `call_indirect` is type-checked at runtime. So basically, two modules need to share g_func_types if one of them imports a funcref or funcref table from the other.

Reply:
I see, that makes sense.  So even for MVP modules that don't include first class funcrefs, we still need to share functype indexes if we want to call_indirect across module boundaries, which, in the general case, we most definitely do.

Reply:
Actually, not that I understand that scope of this index space I do think that "thread local" might be be better approach here.   The reason is that neither tables nor funcrefs are currently shara-able between wasm threads.   So while it makes sense that these type indexes be shared between modules on the same thread, so doesn't make sense to shared them between threads.

So some kind of TLS would probably make more sense here.   In the wasm thread model that we have today, each thread would need to separately instantiate each module.

Reply:
(conversation resolved in the main conversation thread)